src/Microsoft.AspNetCore.Blazor.Browser.JS/src/Rendering/EventForDotNet.ts

```diff
@@ -0,0 +1,46 @@
+﻿export class EventForDotNet<TData extends UIEventArgs> {
+  constructor(public readonly type: EventArgsType, public readonly data: TData) {
+  }
+
+  static fromDOMEvent(event: Event): EventForDotNet<UIEventArgs> {
+    const element = event.target as Element;
+    switch (event.type) {
+      case 'click':
+      case 'mousedown':
+      case 'mouseup':
+        return new EventForDotNet<UIMouseEventArgs>('mouse', { Type: event.type });
+      case 'change': {
+        const targetIsCheckbox = isCheckbox(element);
+        const newValue = targetIsCheckbox ? !!element['checked'] : element['value'];
+        return new EventForDotNet<UIChangeEventArgs>('change', { Type: event.type, Value: newValue });
+      }
+      case 'keypress':
+        return new EventForDotNet<UIKeyboardEventArgs>('keyboard', { Type: event.type, Key: (event as any).key });
+      default:
+        return new EventForDotNet<UIEventArgs>('unknown', { Type: event.type });
+    }
+  }
+}
```

Comment:
Do you think that we'll ever turn the set of payloads here into an open set?

Put another way, would we ever make it possible to write your own `UIEventArgs`-derived class, and wire it up with strongly-typed data? I can't think of a reason why we need it now, but I was wondering if this would become important at some point.

Reply:
I could totally imagine us doing this as some point, either by having some API for plugging in your own event mappers, or perhaps more generally just serializing as much of the JS-side event object as we can and letting developers plug in a `UIEventArgs` subclass on the .NET side to deserialize any matching properties into.

But for now I think we're OK with controlling the set of event mappings ourselves.